File: yardstick_benchmark/source/monitoring/jolokia_get_minecraft_tick.py
#!/usr/bin/env python3

from urllib import request
import json
import time

PERIOD_S = 2.5

def get_tick_durations(old, new):
    assert old is None or len(old) == 100
    assert len(new) == 100

    if old is None:
        return []

    indices_first_new = []
    indices_last_new = []

    for i in range(100):
        j = (i + 1) % 100
        if old[i] == new[i] and old[j] != new[j]:
            indices_first_new.append(j)
        if old[i] != new[i] and old[j] == new[j]:
            indices_last_new.append(i)

    index_first_new = indices_first_new[0]
    index_last_new = indices_last_new[0]

    if len(indices_first_new) != 1 or len(indices_last_new) != 1:
        print("RARE EVENT!")
        maxlen = 0
        for s in indices_first_new:
            for e in indices_last_new:
                d = 0
                if s <= e:
                    d = e - s
                else:
                    d = 100 - s + e + 1
                if d > maxlen:
                    maxlen = d
                    index_first_new = s
                    index_last_new = e


    if index_first_new <= index_last_new:
        return new[index_first_new:index_last_new+1]
    else:
        return new[index_first_new:] + new[:index_last_new+1]


if __name__ == "__main__":
    print("measurement,tick_duration_ms,tick_number,loop_iteration,timestamp_ms,computed_timestamp_ms")

    data_dict = {"type": "read", "mbean": "net.minecraft.server:type=Server", "attribute": "tickTimes", "path": ""}
    data_enc = json.dumps(data_dict).encode('utf-8')
    r = request.Request("http://localhost:8778/jolokia/", data=data_enc)

    prev = None

    t = time.monotonic()

    tick_number = 0
    loop_iteration = 0
    computed_timestamp = None
    prev_tick_duration = None

    while True:
        t += PERIOD_S
        now = time.monotonic()
        time.sleep(t - now)
        with request.urlopen(r) as resp:
            resp_enc = resp.read()
            # 'b{"request":{"mbean":"net.minecraft.server:type=Server","attribute":"tickTimes","type":"read"},"value":[289409,204240,209490,200170,214170,213459,200270,214330,433520,637339,630230,384160,446700,228030,557510,257059,618959,956190,603250,208380,242239,561869,319440,262310,220520,241469,499589,377190,361880,276440,215829,205450,214360,197440,585350,308020,302799,393030,256400,253980,257850,240670,398650,222530,365240,218000,207360,293930,596580,452230,621069,395489,218000,216990,218930,620510,419869,205880,218020,662480,683600,598459,262660,202140,282060,222220,204070,200060,230420,249950,227590,237730,232170,527770,642410,671939,550010,210690,217470,203400,233720,238790,213600,204250,283690,255309,225790,517470,437730,320380,340419,220750,207610,214930,598840,310579,228690,257850,205210,217100],"status":200,"timestamp":1717945225}'
            resp_dict = json.loads(resp_enc.decode('utf-8'))
            curr = resp_dict["value"]
            tick_times = get_tick_durations(prev, curr)
            prev = curr

            for tick_duration in tick_times:
                if computed_timestamp is None:
                    computed_timestamp = now
                else:
                    computed_timestamp += max(50, prev_tick_duration)
                print(f"minecraft_tick_duration,{tick_duration/1000000},{tick_number},{loop_iteration},{now*1000},{computed_timestamp}")
                tick_number += 1
                prev_tick_duration = tick_duration
        loop_iteration += 1

File: yardstick_benchmark/monitoring/jolokia_get_minecraft_tick.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/gammazero/deque"
)

const PERIOD = 1000 * time.Millisecond

type JolokiaResponse struct {
	Value     []int64 `json:"value"`
	Status    int     `json:"status"`
	Timestamp int64   `json:"timestamp"`
}

type Tick struct {
	Duration  time.Duration
	StartTime time.Time
}

func getTickDurations(old, new []int64) []int64 {
	if new == nil || len(new) != 100 {
		panic("new must be length 100")
	}
	if old == nil {
		return []int64{}
	}
	if len(old) != 100 {
		panic("old must be length 100")
	}

	var indicesFirstNew []int
	var indicesLastNew []int

	for i := range 100 {
		j := (i + 1) % 100
		if old[i] == new[i] && old[j] != new[j] {
			indicesFirstNew = append(indicesFirstNew, j)
		}
		if old[i] != new[i] && old[j] == new[j] {
			indicesLastNew = append(indicesLastNew, i)
		}
	}

	// new == old, perhaps the server is paused.
	if len(indicesFirstNew) == 0 || len(indicesLastNew) == 0 {
		return []int64{}
	}

	indexFirstNew := indicesFirstNew[0]
	indexLastNew := indicesLastNew[0]

	if len(indicesFirstNew) != 1 || len(indicesLastNew) != 1 {
		fmt.Println("RARE EVENT!")
		maxlen := 0
		for _, s := range indicesFirstNew {
			for _, e := range indicesLastNew {
				var d int
				if s <= e {
					d = e - s
				} else {
					d = 100 - s + e + 1
				}
				if d > maxlen {
					maxlen = d
					indexFirstNew = s
					indexLastNew = e
				}
			}
		}
	}

	if indexFirstNew <= indexLastNew {
		return new[indexFirstNew : indexLastNew+1]
	}

	return append(new[indexFirstNew:], new[:indexLastNew+1]...)
}

func getDuration(ticks []int64) time.Duration {
	duration := int64(0)
	for _, v := range ticks {
		duration += max(int64(50), v)
	}
	return time.Duration(duration)
}

func main() {
	data := map[string]string{
		"type":      "read",
		"mbean":     "net.minecraft.server:type=Server",
		"attribute": "tickTimes",
		"path":      "",
	}

	dataEnc, err := json.Marshal(data)
	if err != nil {
		panic(err)
	}

	client := &http.Client{}
	url := "http://localhost:8778/jolokia/"

	var prev []int64

	t := time.Now()

	tickNumber := 0
	loopIteration := 0
	var computedTimestamp time.Time
	var prevTickDuration int64

	var tickSlidingWindow deque.Deque[Tick]

	for {
		t = t.Add(PERIOD)
		time.Sleep(time.Until(t))

		req, err := http.NewRequest("POST", url, bytes.NewBuffer(dataEnc))
		if err != nil {
			panic(err)
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			panic(err)
		}

		body, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			panic(err)
		}

		var respObj JolokiaResponse
		err = json.Unmarshal(body, &respObj)
		if err != nil {
			panic(err)
		}
		curr := respObj.Value
		tickTimes := getTickDurations(prev, curr)
		duration := getDuration(tickTimes)
		prev = curr

		now := time.Now()
		// Compute the TPS and print it
		start := now.Add(-duration)
		for _, tickDuration := range tickTimes {
			tickSlidingWindow.PushBack(Tick{Duration: time.Duration(tickDuration), StartTime: start})
			start = start.Add(time.Duration(max(tickDuration, 50)))
		}
		for tickSlidingWindow.Len() > 0 && now.Sub(tickSlidingWindow.Front().StartTime) > 1*time.Second {
			tickSlidingWindow.PopFront()
		}
		tps := tickSlidingWindow.Len()
		fmt.Printf("minecraft_tick tps=%di %d\n", tps, now.UnixNano())

		// Compute the timestamp and duration of every tick and print it
		for _, tickDuration := range tickTimes {
			if computedTimestamp.IsZero() {
				computedTimestamp = now.Add(-duration)
			} else {
				// convert ns to ms for comparison
				prevMs := prevTickDuration / 1_000_000
				sleepMs := max(prevMs, 50)
				computedTimestamp = computedTimestamp.Add(time.Duration(sleepMs) * time.Millisecond)
			}

			fmt.Printf("minecraft_tick tick_duration_ms=%f,tick_number=%di,loop_iteration=%di,timestamp_ms=%di %d\n",
				float64(tickDuration)/1_000_000.0,
				tickNumber,
				loopIteration,
				now.UnixMilli(),
				computedTimestamp.UnixNano(),
			)

			// fmt.Println("measurement,tick_duration_ms,tick_number,loop_iteration,timestamp_ms,computed_timestamp_ms")
			// fmt.Printf(
			// 	"minecraft_tick_duration,%f,%d,%d,%d,%d\n",
			// 	float64(tickDuration)/1_000_000.0,
			// 	tickNumber,
			// 	loopIteration,
			// 	now.UnixMilli(),
			// 	computedTimestamp.UnixMilli(),
			// )

			tickNumber++
			prevTickDuration = tickDuration
		}

		loopIteration++
	}
}
